service/resource/certificate/update.go

```diff
@@ -41,13 +42,38 @@ func (r *Resource) ProcessUpdateState(ctx context.Context, obj, updateState inte
 		return microerror.Mask(err)
 	}
 
-	for _, certificateFile := range updateCertificateFiles {
-		r.logger.Log("debug", fmt.Sprintf("writing certificate: %s", certificateFile.path))
-		if err := afero.WriteFile(r.fs, certificateFile.path, []byte(certificateFile.data), r.certificatePermission); err != nil {
+	// Write the update state certificate files.
+	for _, fileToWrite := range updateCertificateFiles {
+		r.logger.Log("debug", fmt.Sprintf("writing certificate: %s", fileToWrite.path))
+		if err := afero.WriteFile(r.fs, fileToWrite.path, []byte(fileToWrite.data), r.certificatePermission); err != nil {
 			return microerror.Mask(err)
 		}
 	}
 
+	// Remove any unwanted certificate files.
+	fileInfos, err := afero.ReadDir(r.fs, r.certificateDirectory)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	for _, fileInfo := range fileInfos {
+		fileDesired := false
+		filePath := path.Join(r.certificateDirectory, fileInfo.Name())
+
+		for _, updateCertificateFile := range updateCertificateFiles {
+			if filePath == updateCertificateFile.path {
+				fileDesired = true
+			}
+		}
+
+		if !fileDesired {
+			r.logger.Log("debug", fmt.Sprintf("removing certificate: %s", filePath))
+			if err := r.fs.Remove(filePath); err != nil {
+				return microerror.Mask(err)
+			}
+		}
+	}
```

Comment:
IMO most of this should not be done here. Above the update is done nicely. I would expect the same style for deletion. There is just state given to `ProcessUpdateState` and you iterate over it and act on it. All the other state generation magic should actually be done in `GetUpdateState`.

Reply:
For the record we agreed going with this now even though we know it is not that nice. Please create an issue for the refactoring of this.

Reply:
Where is actually the issue now?

Reply:
In the big list in https://github.com/giantswarm/giantswarm/issues/1482